libgalois/src/analytics/matrix_completion/matrix_completion.cpp

```diff
@@ -0,0 +1,534 @@
+  case MatrixCompletionPlan::purdue:
+    return new PurdueStepFunction;
+  case MatrixCompletionPlan::bottou:
+    return new BottouStepFunction;
+  case MatrixCompletionPlan::inverse:
+    return new InverseStepFunction;
+  case MatrixCompletionPlan::bold:
+    return new BoldStepFunction;
+  default:
+    KATANA_DIE("unknown step function");
+  }
+}
+
+
+/*
+ * Generate a number [-1, 1] using node id
+ * for deterministic runs
+ */
+static double
+genVal(uint32_t n) {
+  return 2.0 * ((double)n / (double)RAND_MAX) - 1.0;
+}
+
+double
+sumSquaredError(Graph& graph) {
+  // computing Root Mean Square Error
+  // Assuming only item nodes have edges
+  katana::GAccumulator<double> error;
+
+  katana::do_all(
+      katana::iterate(graph.begin(), graph.begin() + NUM_ITEM_NODES), [&](GNode n) {
+        for (auto ii : graph.edges(n)) {
+          auto dst = graph.GetEdgeDest(ii);
+          double e = predictionError<NodeLatentVector>(graph.GetData<NodeLatentVector>(n),
+              graph.GetData<NodeLatentVector>(dst), graph.GetEdgeData<EdgeWeight>(ii));
+          error += (e * e);
+        }
+      });
+  return error.reduce();
+}
+
+template <typename T, unsigned Size>
+struct ExplicitFiniteChecker {};
+
+template <typename T>
+struct ExplicitFiniteChecker<T, 4U> {
+  static_assert(std::numeric_limits<T>::is_iec559, "Need IEEE floating point");
+  bool isFinite(T v) {
+    union {
+      T value;
+      uint32_t bits;
+    } a = {v};
+    if (a.bits == 0x7F800000) {
+      return false;  // +inf
+    } else if (a.bits == 0xFF800000) {
+      return false;  // -inf
+    } else if (a.bits >= 0x7F800001 && a.bits <= 0x7FBFFFFF) {
+      return false;  // signaling NaN
+    } else if (a.bits >= 0xFF800001 && a.bits <= 0xFFBFFFFF) {
+      return false;  // signaling NaN
+    } else if (a.bits >= 0x7FC00000 && a.bits <= 0x7FFFFFFF) {
+      return false;  // quiet NaN
+    } else if (a.bits >= 0xFFC00000 && a.bits <= 0xFFFFFFFF) {
+      return false;  // quiet NaN
+    }
+    return true;
+  }
+};
+
+template <typename T>
+struct ExplicitFiniteChecker<T, 8U> {
+  static_assert(std::numeric_limits<T>::is_iec559, "Need IEEE floating point");
+  bool isFinite(T v) {
+    union {
+      T value;
+      uint64_t bits;
+    } a = {v};
+    if (a.bits == 0x7FF0000000000000) {
+      return false;  // +inf
+    } else if (a.bits == 0xFFF0000000000000) {
+      return false;  // -inf
+    } else if (a.bits >= 0x7FF0000000000001 && a.bits <= 0x7FF7FFFFFFFFFFFF) {
+      return false;  // signaling NaN
+    } else if (a.bits >= 0xFFF0000000000001 && a.bits <= 0xFFF7FFFFFFFFFFFF) {
+      return false;  // signaling NaN
+    } else if (a.bits >= 0x7FF8000000000000 && a.bits <= 0x7FFFFFFFFFFFFFFF) {
+      return false;  // quiet NaN
+    } else if (a.bits >= 0xFFF8000000000000 && a.bits <= 0xFFFFFFFFFFFFFFFF) {
+      return false;  // quiet NaN
+    }
+    return true;
+  }
+};
+
+template <typename T>
+bool
+isFinite(T v) {
```

Comment:
Can we take out all the helper functions in a separate file similar to "libgalois/include/katana/analytics/ClusteringImplementationBase.h"

Reply:
I moved the helper functions that are not dependent on MatrixCompletionPlan to MatrixCompletionImplementationBase.h, similar to ClusteringImplementationBase.h and BfsSsspImplementationBase.h